internal/buildengine/engine.go

```diff
@@ -781,3 +845,81 @@ func (e *Engine) gatherSchemas(
 
 	return nil
 }
+
+func (e *Engine) newModuleMeta(ctx context.Context, config moduleconfig.ModuleConfig, projectPath string) (moduleMeta, error) {
+	plugin, err := PluginFromConfig(ctx, config, projectPath)
+	if err != nil {
+		return moduleMeta{}, fmt.Errorf("could not create plugin for %s: %w", config.Module, err)
+	}
+	events := make(chan PluginEvent, 64)
+	plugin.Updates().Subscribe(events)
+
+	// pass on plugin events to the main event channel
+	// make sure we do not pass on nil (chan closure) events
+	go func() {
+		for {
+			select {
+			case event := <-events:
+				if event == nil {
+					// chan closed
+					return
+				}
+				e.pluginEvents <- event
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return moduleMeta{
+		module: Module{
+			Config:       config,
+			Dependencies: []string{},
+		},
+		plugin: plugin,
+		events: events,
+	}, nil
+}
+
+// listenForBuildUpdates listens for adhoc build updates and reports them to the listener.
+// These happen when a plugin for a module detects a change and automatically rebuilds.
+func (e *Engine) listenForBuildUpdates(originalCtx context.Context) {
+	for {
+		select {
+		case event := <-e.pluginEvents:
+			logger := log.FromContext(originalCtx).Module(event.ModuleName()).Scope("build")
```

Comment:
Maybe create the logger once at the top?

Reply:
The logger needs to be scoped to the module, but the loop is for all modules. So it needs to be done per event